packages/modelserver-client/src/utils/type-util.ts

```diff
@@ -145,3 +146,172 @@ export function asModelArray(object: unknown): Model[] {
+    return traverse(object, (target, props) => props.includes(name) ? true : undefined);
+}
+
+/**
+ * Copy an object graph for the sole purpose of writing over the wire in a JSON format.
+ *
+ * @param object the object graph to copy
+ * @param format the JSON format in which to copy it
+ * @returns the copied object graph
+ */
+function copy(object: any, format: JsonFormat): any {
+    // A map of original to copy
+    const copies: Map<any, any> = new Map();
+
+    const copier = (target: any): void => {
+        const theCopy = { ...target };
+
+        if (format === 'json' && '$type' in theCopy) {
+            theCopy.eClass = theCopy.$type;
+            delete theCopy.$type;
+        } else if (format === 'json-v2' && 'eClass' in theCopy) {
+            theCopy.$type = theCopy.eClass;
+            delete theCopy.eClass;
+        }
+
+        copies.set(target, theCopy);
+    };
+
+    const referencer = (target: any, props: string[]): void => {
+        const copyGetter = (o: any): any => copies.get(o) ?? o;
+
+        // We want to traverse all properties of object type; already guarded via `getOwnPropertyNames`.
+        // eslint-disable-next-line guard-for-in
+        for (const prop of props) {
+            const value = target[prop];
+
+            if (isNonEmptyObjectArray(value)) {
+                target[prop] = value.map(copyGetter);
+            } else if (Array.isArray(value)) {
+                // It's an array of non-object data types
+                target[prop] = [...value];
+            } else if (typeof value === 'object') {
+                target[prop] = copyGetter(value);
+            }
+        }
+    };
+
+    // Step one: copy everything
+    traverse(object, copier);
+    const result = copies.get(object);
+
+    // Step two: then rewrite cross-references
+    traverse(result, referencer);
+
+    return result;
+}
+
+function traverse<T>(object: any, fn: (target: any, props: string[]) => T | undefined, visited = new Set()): boolean {
+    if (visited.has(object)) {
+        return false;
+    }
+    visited.add(object);
+
+    const ownProps = Object.getOwnPropertyNames(object);
+
+    const result = fn(object, ownProps);
+    if (result !== undefined) {
+        return true;
+    }
+
+    // We want to traverse all properties of object type; already guarded via `getOwnPropertyNames`.
+    // eslint-disable-next-line guard-for-in
+    for (const prop of ownProps) {
+        const value = object[prop];
+
+        if (isNonEmptyObjectArray(value) && value.some(element => traverse(element, fn, visited))) {
+            return true;
+        }
+
+        if (typeof value === 'object' && traverse(value, fn, visited)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+function isNonEmptyObjectArray(value: any): value is AnyObject[] {
+    // TODO: More correct would be value.every(e => typeof e === 'object' && !Array.isArray(value[0]))
```

Comment:
@cdamus ?

Reply:
I emulated examples found elsewhere in the EMF.Cloud universe, but then added my own commentary. I'm happy either to remove the comment or to implement its suggestion, except that the suggestion I think is actually wrong: it doesn't actually reject an empty array because `every` is always `true` for empty arrays.

Reply:
So in that case, I would suggest to simply remove the comment for now unless we actually run into problems with it.
@CamilleLetavernier , @cdamus Could either of you adapt the PR or accept the suggestion?
```suggestion
```